### PR TITLE
add collection search extension

### PR DESCRIPTION
**Pull request by hrodmn**
**Related Issue(s):**

- #135 

**Description:**
Adds the collection search extension by pointing the `all_collections` method at the pgstac function `collection_search` instead of `all_collections`.

I have not tested to see if paging is implemented yet, need to upload a bunch of collections to my local db I guess! :sweat_smile:

**PR Checklist:**

- [x] `pre-commit` hooks pass locally
- [x] Tests pass (run `make test`)
- [ ] Documentation has been updated to reflect changes, if applicable, and docs build successfully (run `make docs`)
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/stac-fastapi/blob/main/CHANGES.md).

### CODE REVIEW

File: stac_fastapi/pgstac/core.py [hunk 1]
```diff
@@ -39,17 +39,113 @@
...
 
-    async def all_collections(self, request: Request, **kwargs) -> Collections:
-        """Read all collections from the database."""
+    async def all_collections(  # noqa: C901
+        self,
+        request: Request,
+        bbox: Optional[BBox] = None,
+        datetime: Optional[DateTimeType] = None,
+        limit: Optional[int] = None,
+        # Extensions
```
Comment (vincentsarago) [inline stac_fastapi/pgstac/core.py:48]: you can move the `Extensions` comment before `bbox`

File: stac_fastapi/pgstac/core.py [hunk 2]
```diff
@@ -39,17 +39,113 @@
...
-                SELECT * FROM all_collections();
-                """
-            )
+        settings: Settings = request.app.state.settings
+
+        if search_request.datetime:
+            search_request.datetime = format_datetime_range(search_request.datetime)
+
+        search_request.conf = search_request.conf or {}
+        search_request.conf["nohydrate"] = settings.use_api_hydrate
```
Comment (vincentsarago) [inline stac_fastapi/pgstac/core.py:124]: is there hydratation for collections?

File: stac_fastapi/pgstac/core.py [hunk 3]
```diff
@@ -39,17 +39,113 @@
...
+            Collections which match the search criteria, returns all
+            collections by default.
+        """
+        query_params = str(request.query_params)
+
+        # Kludgy fix because using factory does not allow alias for filter-lang
+        if filter_lang is None:
+            match = re.search(r"filter-lang=([a-z0-9-]+)", query_params, re.IGNORECASE)
+            if match:
+                filter_lang = match.group(1)
```
Comment (vincentsarago) [inline stac_fastapi/pgstac/core.py:71]: I don't think this is true anymore

File: stac_fastapi/pgstac/core.py [hunk 4]
```diff
@@ -39,17 +39,113 @@
...
-        async with request.app.state.get_connection(request, "r") as conn:
-            collections = await conn.fetchval(
-                """
-                SELECT * FROM all_collections();
-                """
-            )
+        settings: Settings = request.app.state.settings
+
+        if search_request.datetime:
+            search_request.datetime = format_datetime_range(search_request.datetime)
```
Comment (vincentsarago) [inline stac_fastapi/pgstac/core.py:121]: it seems that this will be done twice 🤔

File: stac_fastapi/pgstac/core.py [hunk 5]
```diff
@@ -39,17 +39,113 @@
...
+        return await self._collection_search_base(search_request, request=request)
+
+    async def _collection_search_base(  # noqa: C901
+        self,
+        search_request: PgstacSearch,
+        request: Request,
+    ) -> Collections:
+        """Cross catalog search (POST).
+
+        Called with `POST /search`.
```
Comment (vincentsarago) [inline stac_fastapi/pgstac/core.py:107]:
```suggestion
        Called with `GET /collections`.
```

File: stac_fastapi/pgstac/core.py [hunk 6]
```diff
@@ -39,17 +39,113 @@
...
+            datetime=datetime,
+            fields=fields,
+            sortby=sortby,
+            filter=filter,
+            filter_lang=filter_lang,
+        )
+
+        # Do the request
+        try:
+            search_request = self.post_request_model(**clean)
```
Comment (vincentsarago) [inline stac_fastapi/pgstac/core.py:92]: 🤔 😬 the issue is that here we don't have to the model used, we should add a `collection_get_request_model` attribute to the client or find a way to avoid this

File: stac_fastapi/pgstac/core.py (hunk 3 above)
Comment (hrodmn) [reply stac_fastapi/pgstac/core.py:71]: Yes it looks like you are right!

https://github.com/stac-utils/stac-fastapi/blob/e376e3000c0231d3bbf8e12a8f7a9a254067b3c9/stac_fastapi/extensions/stac_fastapi/extensions/core/filter/request.py#L37-L43

File: stac_fastapi/pgstac/core.py (hunk 6 above)
Comment (hrodmn) [reply stac_fastapi/pgstac/core.py:92]: If I understand the problem correctly, we are just trying to dump the args into a json that pgstac will accept, but we are using a model for item search that is defined in `stac_pydantic`. Does this mean we want to add a new attribute to the `CoreCrudClient` that is equivalent to `CollectionSearchExtension().GET`?

Comment (vincentsarago) [reply stac_fastapi/pgstac/core.py:92]: it's kinda weird because we will pass the Client to the CollectionSearch extension, which kinda result in a circular dependency, but that's how it's done for the `search post model` https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/types/stac_fastapi/types/core.py#L346

> Does this mean we want to add a new attribute to the CoreCrudClient that is equivalent to CollectionSearchExtension().GET?

But yes I think we will have to do that!

**Comment by m-mohr** [comment]
A queryables link seems to be missing in GET /collections

**Comment by vincentsarago** [comment]
> Great work. A queryables link seems to be missing in GET /collections, right?

@m-mohr correct, we will get the changes we just pushed to main from #89

**Comment by m-mohr** [comment]
@vincentsarago #89 didn't solve it for Collection Search, that was only for Item Search. So getting #89 doesn't change anything with regards to my request.

File: stac_fastapi/pgstac/core.py (hunk 6 above)
Comment (hrodmn) [reply stac_fastapi/pgstac/core.py:92]: I tried adding a `collections_get_request_model` attribute based on the `BaseCollectionSearchGetRequest` class but it doesn't work because that class is not a pydantic model. It looks like the `post_request_model` is a `BaseSearchPostRequest` which inherits from `stac_pydantic.api.Search` (which is based on a BaseModel), but `BaseSearchGetRequest` is based on the `APIRequest` class which does not inherit from a pydantic model and thus has no `model_dump_json()` method.

Here is what I tried:
```python

@attr.s
class CoreCrudClient(AsyncBaseCoreClient):
    """Client for core endpoints defined by stac."""

    collections_get_request_model: BaseCollectionSearchGetRequest = attr.ib(
        default=BaseCollectionSearchGetRequest
    )

    async def all_collections(  # noqa: C901
        self,
        request: Request,
        # Extensions
        bbox: Optional[BBox] = None,
        datetime: Optional[DateTimeType] = None,
        limit: Optional[int] = None,
        query: Optional[str] = None,
        token: Optional[str] = None,
        fields: Optional[List[str]] = None,
        sortby: Optional[str] = None,
        filter: Optional[str] = None,
        filter_lang: Optional[str] = None,
        **kwargs,
    ) -> Collections:
        """Cross catalog search (GET).

        Called with `GET /collections`.

        Returns:
            Collections which match the search criteria, returns all
            collections by default.
        """

        # Parse request parameters
        base_args = {
            "bbox": bbox,
            "limit": limit,
            "token": token,
            "query": orjson.loads(unquote_plus(query)) if query else query,
        }

        clean = clean_search_args(
            base_args=base_args,
            datetime=datetime,
            fields=fields,
            sortby=sortby,
            filter=filter,
            filter_lang=filter_lang,
        )

        # Do the request
        try:
            search_request = self.collections_get_request_model(**clean)
        except ValidationError as e:
            raise HTTPException(
                status_code=400, detail=f"Invalid parameters provided {e}"
            ) from e

        return await self._collection_search_base(search_request, request=request)


```

So I think I will need to get the json for the pgstac query using a different approach.

File: stac_fastapi/pgstac/core.py (hunk 2 above)
Comment (hrodmn) [reply stac_fastapi/pgstac/core.py]: I don't think so

File: stac_fastapi/pgstac/core.py (hunk 6 above)
Comment (hrodmn) [reply stac_fastapi/pgstac/core.py]: I got this working by using `BaseCollectionSearchPostRequest` (instead of `BaseCollectionSearchGetRequest`) which follows the pattern used in the item search.

**Comment by hrodmn** [comment]
> > Great work. A queryables link seems to be missing in GET /collections, right?
> 
> @m-mohr correct, we will get the changes we just pushed to main from #89

I rebased on main so this branch should be up-to-date with the queryables link now.

File: stac_fastapi/pgstac/core.py (hunk 6 above)
Comment (vincentsarago) [reply stac_fastapi/pgstac/core.py]: for get request I think we should use the GET model or it might not work properly! 

Maybe I was wrong before and looking at the code, I think we can use `PgstacSearch` at line 91 

or we could create another pydantic model which match what `pgstac.collection_search` expect

Comment (hrodmn) [reply stac_fastapi/pgstac/core.py]: Ah, I think I see the path. I probably need to use `create_request_model` to make sure the appropriate args get added for the extensions. However, without creating a new pydantic model, I would still need to use the `POST` model because the search functionality relies on the pydantic model. That's probably something that belongs in stac-fastapi, right? Is there a particular reason the `GET` request models are a generic class rather than a pydantic model in stac-fastapi?

Comment (vincentsarago) [reply stac_fastapi/pgstac/core.py]: >  Is there a particular reason the GET request models are a generic class rather than a pydantic model in stac-fastapi?

yes because pydantic model are interpreted as body parameter for endpoints

Comment (hrodmn) [reply stac_fastapi/pgstac/core.py]: Thanks for your patience while I figure this out. I think I need to double back and make sure I understand how all of these models are getting used. Upon closer examination, it looks like stac-fastapi-pgstac is just using the `AsyncBaseCoreClient.post_request_model` for all of the POST requests:
e.g.
https://github.com/stac-utils/stac-fastapi-pgstac/blob/cb08b0f6f1b11643632071113e10921501da3808/stac_fastapi/pgstac/core.py#L324

We provide a `search_post_request_model` to `StacAPI`:
https://github.com/stac-utils/stac-fastapi-pgstac/blob/cb08b0f6f1b11643632071113e10921501da3808/stac_fastapi/pgstac/app.py#L67-L78

But I can't see how `search_post_request_model` is getting used by anything right now.

Comment (hrodmn) [reply stac_fastapi/pgstac/core.py:84]: I guess `search_get_request_model` and `search_post_request_model` are a dead end. We pass the `post_request_model` that is based on `PgstacSearch` to the `CoreCrudClient` and that is what gets used to generate inputs for all of the `pgstac` search functions. 

I am pretty sure we can use it in the same way for the `/collections` GET method like we do in the `/collections/{collection_id}/items` GET method.

**Comment by hrodmn** [comment]
@vincentsarago are you worried about the possibility of inconsistencies between the arguments in `PgstacSearch` and the collection-search GET request model? I believe the collection-search GET request should have a subset of the parameters from item-level search and never any extra parameters. If I pass a `collections_get_request_model` that should limit the args that get passed to `PgstacSearch` in the `all_collections` method to the ones that are available in the collection-search endpoint with all of the enabled extensions.

I'm sorry about all of the force pushes, I had to clean some things up in order to make the changes mergeable with main :sweat_smile:

File: stac_fastapi/pgstac/app.py [hunk 7]
```diff
@@ -64,17 +83,30 @@
 else:
     items_get_request_model = ItemCollectionUri
 
-post_request_model = create_post_request_model(extensions, base_model=PgstacSearch)
-get_request_model = create_get_request_model(extensions)
+if any(isinstance(ext, CollectionSearchExtension) for ext in collection_extensions):
+    collections_get_request_model = create_request_model(
+        model_name="CollectionsGetRequest",
+        base_model=BaseCollectionSearchGetRequest,
+        extensions=extensions,
```
Comment (hrodmn) [inline stac_fastapi/pgstac/app.py:90]: Enable args from other extensions in the `/collections` GET request model

File: stac_fastapi/pgstac/core.py [hunk 8]
```diff
@@ -39,17 +38,100 @@
...
 
-    async def all_collections(self, request: Request, **kwargs) -> Collections:
-        """Read all collections from the database."""
+    async def all_collections(  # noqa: C901
+        self,
+        request: Request,
+        # Extensions
+        bbox: Optional[BBox] = None,
+        datetime: Optional[DateTimeType] = None,
+        limit: Optional[int] = None,
```
Comment (alukach) [inline stac_fastapi/pgstac/core.py:47]: Thoughts from call: consider adding `id` to the endpoint, despite it being omitted from the [Collection Search Extension](https://github.com/stac-api-extensions/collection-search?tab=readme-ov-file#query-parameters-and-fields).

File: stac_fastapi/pgstac/app.py (hunk 7 above)
Comment (vincentsarago) [reply stac_fastapi/pgstac/app.py:90]: should `extensions` should be `collection_extensions` ?

Comment (vincentsarago) [reply stac_fastapi/pgstac/app.py:90]: also I don't think we need to `test` with isinstance, just a `if collection_extensions:` should be enough

Comment (hrodmn) [reply stac_fastapi/pgstac/app.py:90]: It needs to be `extensions` because want to add the parameters from other extensions (like `Filter`, `Sort`) to the collections GET request if those extensions are enabled. It's confusing but this only gets run **if** the collection search extension is enabled.

Comment (vincentsarago) [reply stac_fastapi/pgstac/app.py:90]: so `collection_extensions` is not really used then? (sorry I'm trying to catch back after two weeks away from stac-fastapi-* 😓)

File: stac_fastapi/pgstac/core.py (hunk 8 above)
Comment (m-mohr) [reply stac_fastapi/pgstac/core.py:47]: Isn't that the same as just calling /collections/:id ?

File: stac_fastapi/pgstac/app.py (hunk 7 above)
Comment (vincentsarago) [reply stac_fastapi/pgstac/app.py:90]: ok so:
- we want to check if `collection_search` is enabled in `ENABLED_EXTENSIONS` env. 
- if enable, we need to construct a model based on `BaseCollectionSearchGetRequest` which we can extent with the other extensions

makes sense 👍 , maybe add some docstring reflecting ☝️ 

I feel splitting the to `extensions` and `collection_extensions` might not be needed. I think I'm the one who proposed it but maybe it's useless

Comment (vincentsarago) [reply stac_fastapi/pgstac/app.py:90]: well looking at https://github.com/stac-utils/stac-fastapi/blob/4979a89b4dc1534645388f182046955986fee669/stac_fastapi/extensions/stac_fastapi/extensions/core/collection_search/collection_search.py#L20-L72 

it's a bit more complex, and I don't think we can simply add extension to the collection-search. Technically we would also need to change the conformances list if `sort` or `filter` were added to the collection-search

IMO, the application should implement a minimal solution (simple collection-search) but we can focus on the documentation to explain how to extend the collection-search to allow extensions

Comment (hrodmn) [reply stac_fastapi/pgstac/app.py:90]: I think adding extensions to the collection-search extension will be really useful (especially `free-text` and `filter`). I'll come up with something that can detect which conformance classes to add based on the enabled extensions. I agree that it is a little messy, so if this is another case where you would want to have users enable it via custom runtimes I guess that would work. I just think robust collection-search would be a really nice feature to have on by default!

Comment (hrodmn) [reply stac_fastapi/pgstac/app.py:90]: That looks like a much cleaner solution! I'll update this PR to skip adding extensions to the collection-search extension then follow up with a PR in stac-fastapi with this `classmethod`.

File: stac_fastapi/pgstac/core.py (hunk 8 above)
Comment (hrodmn) [reply stac_fastapi/pgstac/core.py:47]: The use-case for including an `ids` parameter would be to limit the scope of a search in the context of scoped authentication for a STAC API, but we discussed some more and it probably makes more sense to use the `filter` extension for injecting scope limits in a search request.

Comment (alukach) [reply stac_fastapi/pgstac/core.py:47]: Yes, `ids` would be a better parameter name than `id` to evoke the ability to provide multiple as a filter. And for our specific needs today, we will make use of `filter` as @hrodmn said.  I believe it was @bitner who suggested adding the ability to filter by ID parameter, I'll let him weigh in on whether we should go without.

Comment (bitner) [reply stac_fastapi/pgstac/core.py:47]: I think it makes sense just to be parallel to the items spec (and yes, "ids" plural which is how it works in items)

Comment (m-mohr) [reply stac_fastapi/pgstac/core.py:47]: If you add it, I'd recommend to add a separate conformance class for it so that clients actually know whether it's supported or not.

PS: ids is not included in collection search as we just inherit from OGC API - Records, which doesn't have it. It's orthogonal to how ids is not part of OGC API - Features for items. ids is a STAC-specific thing.

**Comment by hrodmn** [comment]
@vincentsarago I think we need a new release of stac-fastapi with the changes from https://github.com/stac-utils/stac-fastapi/pull/745 in order to properly configure the collection-search extension with the other configured extensions.

File: stac_fastapi/pgstac/app.py (hunk 7 above)
Comment (hrodmn) [reply stac_fastapi/pgstac/app.py]: I bumped the stac-fastapi requirement to `3.0.2` so we can use the new method - thanks for adding that @vincentsarago!

File: stac_fastapi/pgstac/app.py [hunk 9]
```diff
@@ -64,17 +71,33 @@
...
+    collections_get_request_model = collection_search_extension.GET
+else:
+    collection_search_extension = None
+    collections_get_request_model = EmptyRequest
+
 post_request_model = create_post_request_model(extensions, base_model=PgstacSearch)
 get_request_model = create_get_request_model(extensions)
 
+# will only use parameters defined in collections_get_request_model
+collection_search_model = create_post_request_model(extensions, base_model=PgstacSearch)
```
Comment (vincentsarago) [inline stac_fastapi/pgstac/app.py:87]: it's a bit of a mess having `collections_get_request_model` and `collection_search_model`

maybe we can make the variable names a bit clearer

Comment (hrodmn) [reply stac_fastapi/pgstac/app.py]: I'm not sure why I didn't try this earlier, but I think recycling the `collections_get_request_model` in the `CoreCrudClient` is the cleanest way. We lose the `model_dump_json` and `alias` functionality from the other pydantic models, but it is less confusing than trying to add another model here.

File: stac_fastapi/pgstac/core.py [hunk 10]
```diff
@@ -83,7 +84,8 @@ async def all_collections(  # noqa: C901
 
         # Do the request
         try:
-            search_request = self.collection_request_model(**clean)
+            search_request = self.collections_get_request_model(**clean)
+            print(search_request)
```
Comment (vincentsarago) [inline stac_fastapi/pgstac/core.py:88]: 👀

**Comment by hrodmn** [comment]
It looks like I am having some `datetime` json issues in Python 3.10, I'll figure out a fix.

File: stac_fastapi/pgstac/app.py [hunk 11]
```diff
@@ -64,17 +71,30 @@
...
+    collections_get_request_model = EmptyRequest
+
 post_request_model = create_post_request_model(extensions, base_model=PgstacSearch)
 get_request_model = create_get_request_model(extensions)
 
 api = StacApi(
     settings=settings,
-    extensions=extensions,
-    client=CoreCrudClient(post_request_model=post_request_model),  # type: ignore
+    extensions=extensions + [collection_search_extension],
```
Comment (vincentsarago) [inline stac_fastapi/pgstac/app.py:88]: when there is no `collection_search` extension we will have `[None]`. I think we should avoid this

File: stac_fastapi/pgstac/core.py [hunk 12]
```diff
@@ -39,17 +39,66 @@
...
+
         async with request.app.state.get_connection(request, "r") as conn:
-            collections = await conn.fetchval(
-                """
-                SELECT * FROM all_collections();
+            q, p = render(
                 """
+                SELECT * FROM collection_search(:req::text::jsonb);
+                """,
+                req=json.dumps(clean_args),
```
Comment (vincentsarago) [inline stac_fastapi/pgstac/core.py:89]: we don't need to re-validate the query, this is already done!

**Review by bitner — APPROVED** [review]

File: stac_fastapi/pgstac/app.py [hunk 13]
```diff
@@ -64,17 +71,29 @@
...
 
+if "collection_search" in _enabled_extensions:
+    collection_search_extension = CollectionSearchExtension.from_extensions(
+        extensions=extensions
+    )
+    collections_get_request_model = collection_search_extension.GET
+else:
+    collection_search_extension = None
+    collections_get_request_model = EmptyRequest
+
```
Comment (alukach) [inline stac_fastapi/pgstac/app.py:82]: This may just be a _me_ thing, but I find it a bit hard to follow variables being created within `if ... else` scope.  Inverting these may help legibility.
```py
enabled_extensions = (
    os.getenv("ENABLED_EXTENSIONS").split(",")
    if "ENABLED_EXTENSIONS" in os.environ
    else extensions_map.keys() + ["collection_search"]
)
extensions = [
    extension for key, extension in extensions_map.items() if key in enabled_extensions
]

items_get_request_model = (
    create_request_model(
        model_name="ItemCollectionUri",
        base_model=ItemCollectionUri,
        mixins=[TokenPaginationExtension().GET],
        request_type="GET",
    )
    if any(isinstance(ext, TokenPaginationExtension) for ext in extensions)
    else ItemCollectionUri
)

collection_search_extension = (
    CollectionSearchExtension.from_extensions(extensions)
    if "collection_search" in enabled_extensions
    else None
)
collections_get_request_model = (
    collection_search_extension.GET if collection_search_extension else EmptyRequest
)
```

**Review by alukach — APPROVED** [review]
LGTM. Left some very non-critical suggestions.

File: stac_fastapi/pgstac/core.py (hunk 8 above)
Comment (hrodmn) [reply stac_fastapi/pgstac/core.py:48]: Unless someone else feels strongly, I think we can leave `ids` out since it is not in the extension spec and if a user knows the collection IDs they could just make several calls to `/collections/{id}`.

**Comment by hrodmn** [comment]
@vincentsarago thanks for your help getting this cleaned up, I think everything is looking good. I added some commits based on @alukach's suggestions. The only outstanding conversation is about the `ids` parameter which might be occasionally convenient but I think it would be ok to proceed without it since it's not in the collection search extension spec.
